docs/specs/snp-genotyping-vector.md

```diff
@@ -0,0 +1,185 @@
+    -XA RMSMappingQuality \
+    -XA Coverage \
+    -XA ExcessHet \
+    -XA InbreedingCoeff \
+    -XA MappingQualityZero \
+    -XA HaplotypeScore \
+    -XA SpanningDeletions \
+    -XA FisherStrand \
+    -XA StrandOddsRatio \
+    -XA ChromosomeCounts \
+    -XA BaseQualityRankSumTest \
+    -XA MappingQualityRankSumTest \
+    -XA QualByDepth \
+    -XA ReadPosRankSumTest
+```
+
+Notes: 
+ 
+1. Because we are genotyping a single sample against given alleles, we
+   don't need many of the variant annotations that we would usually
+   ask for when doing discovery. Reducing the number of annotations
+   should reduce file size somewhat.
+
+2. I've set the '-contamination' option to zero (0.0) here rather than
+   0.05 we've used previously because I didn't realise that this
+   option down-samples only alternate alleles (I thought it
+   downsampled both alt and ref alleles), and so setting contamination
+   > 0 will create bias towards reference allele. We will deal with
+   contamination separately.
+
+3. When genotyping given alleles, GATK ignores any variant in the
+   known variants VCF that is not marked as PASS in the FILTER
+   field. However, we want to genotype at all variants in the known
+   variants VCF, regardless of what is in the FILTER column. Therefore
+   the known variants VCF must be created setting FILTER to PASS for
+   all variants to get this to work as desired.
+
+
+### Step: VCF to Zarr conversion
+
+Convert the VCF to Zarr format via the scikit-allel
+[vcf_to_zarr()](https://scikit-allel.readthedocs.io/en/stable/io.html#allel.vcf_to_zarr)
+function. Note the following implementation details:
+
+* Only the ``variants/MQ``, ``calldata/GT``, ``calldata/GQ`` and
+  ``calldata/AD`` fields are required in the output.
+
+* Data types should be as follows: ``variants/MQ``: i1 (single byte
+  signed integer), ``calldata/GT``: i1; ``calldata/GQ``: i1;
+  ``calldata/AD``: i2
+
+* Compressor should be ``zarr.Blosc(cname='zstd', clevel=5,
+  shuffle=-1)``
+
+* Chunk length should be ``2**18``
+
+* ``alt_number`` parameter should be 3 (fix number of alt alleles)
+
+* Data should be grouped by contig (chromosome). This means calling
+  ``vcf_to_zarr()`` once for each contig in the reference genome,
+  specifying the name of the contig as the value of the ``region`` and
+  ``group`` parameters.
+
+Here is an example of the appropriate function call in Python for a
+single contig (this should be called for all contigs in the genome):
+
+```python
+import sys
+import allel
+import zarr
+
+
+def build_sample_zarr(input_path, output_path, contig):
+    allel.vcf_to_zarr(
+        input=input_path,
+        output=output_path,
+        group=contig,
+        region=contig,
+        compressor=zarr.Blosc(cname='zstd', clevel=5, shuffle=0),
+        overwrite=True,
+        fields=['calldata/GT', 'calldata/GQ', 'calldata/AD', 'variants/MQ'],
+        types={'calldata/GT': 'i1',
+               'calldata/GQ': 'i1',
```

Comment:
I think explicitly setting `GQ` and `MQ` as integers is a change from the previous/current version and could be noted in the change log.

Reply:
The theoretical maximum of `MQ` (from SAM format specs) is 255- therefore we need `i2`:
`GQ` is capped at 99, so `i1` is fine.

Also, might be worth noting here that MQ is the RMS mapping quality of all reads contributing to genotype call.

Reply:
Also, `GQ`/`MQ` can be missing. Unsure how `allel` handles this @alimanfoo ? Should be coded as `-1` in int output.

Reply:
> I think explicitly setting `GQ` and `MQ` as integers is a change from the previous/current version and could be noted in the change log.

The [previous version](https://github.com/malariagen/vector-ops/blob/a1227e1ae0dcdf950ecc633cb63c4da44711da66/specs/genotyping.md) did specify a dtype for GQ but was actually missing the MQ field completely, although we did include MQ in the implementation. I've added a note.

Reply:
> The theoretical maximum of `MQ` (from SAM format specs) is 255- therefore we need `i2`:

I'm pretty sure that the maximum MQ from bwa is 60, so 'i1' should be fine.

Reply:
> Also, `GQ`/`MQ` can be missing. Unsure how `allel` handles this @alimanfoo ? Should be coded as `-1` in int output.

Yes, all integer fields should get -1 if missing.

Reply:
Cool- just checked all phase 3 zip zarrs. `70` is the max observed.

Reply:
I believe the /variants/MQ field in the AG1K phase 3 zarrs were float32.  UnifiedGenotyper sets MQ annotation as the root mean square of alignment mapping quality as @hardingnj mentioned.  Just wanted to confirm - are we OK with losing the decimal precision for the /variants/MQ field in the zarrs going forward?

Reply:
> Cool- just checked all phase 3 zip zarrs. `70` is the max observed.

Thanks for doing that Nick.

Reply:
> I believe the /variants/MQ field in the AG1K phase 3 zarrs were float32. UnifiedGenotyper sets MQ annotation as the root mean square of alignment mapping quality as @hardingnj mentioned. Just wanted to confirm - are we OK with losing the decimal precision for the /variants/MQ field in the zarrs going forward?

Ah OK. Yes should be fine as int8, definitely don't need the decimal precision.

Reply:
@alimanfoo @hardingnj  Sorry, I realize I left out some information.  To clarify, the MQ annotation is the root mean square **+ standard deviation**.  (According to:  https://gatk.broadinstitute.org/hc/en-us/articles/360037591751-RMSMappingQuality).  We'll lose information if we drop from float32 to int8.  Is this still OK?

Reply:
Hi @tnguyensanger, bottom line is that range of values of MQ is 0-70, decimal precision is not needed for any analysis we do.

MQ is calculated as root mean square, which means square mapping qualities from all reads, sum the quares, divide by number of reads, take the root.

The "statistical note" in the page you linked says that RMS is "equivalent to the mean of the mapping qualities plus the standard deviation of the mapping qualities". That may be true, I haven't checked. In either case, it's just a point of statistical interest. Doesn't change the fact that we don't need the decimal precision.

Hth.

Reply:
Sounds good to me.